imagecraft/pack/chroot.py

```diff
@@ -126,28 +124,34 @@ def __init__(self, *, path: Path, mounts: list[Mount]) -> None:
         self.path = path
         self.mounts = mounts
 
-    def _setup_chroot(self) -> None:
+    def _setup(self) -> None:
         """Chroot environment preparation."""
-        # Some images (such as cloudimgs) symlink ``/etc/resolv.conf`` to
-        # ``/run/systemd/resolve/stub-resolv.conf``. We want resolv.conf to be
-        # a regular file to bind-mount the host resolver configuration on.
-        #
-        # There's no need to restore the file to its original condition because
-        # this operation happens on a temporary filesystem layer.
         logger.debug("setup chroot: %r", self.path)
 
         for entry in self.mounts:
             entry.mount(base_path=self.path)
 
         logger.debug("chroot setup complete")
 
-    def _cleanup_chroot(self) -> None:
+    def _cleanup(self) -> None:
         """Chroot environment cleanup."""
+        umount_errors: list[str] = []
         logger.debug("cleanup chroot: %r", self.path)
         for entry in reversed(self.mounts):
-            entry.umount()
+            try:
+                entry.umount()
+            except subprocess.CalledProcessError as err:  # noqa: PERF203
+                msg = str(err)
+                if err.stderr:
+                    msg += f" ({err.stderr.strip()!s})"
+                umount_errors.append(msg)
+
+        if umount_errors:
+            raise errors.ChrootExecutionError(
+                message="Failed to clean chroot", details="\n".join(umount_errors)
+            )
 
-    def chroot(
+    def execute(
```

Comment:
It's very out-of-scope here, but having a context manager for the chroot and having this method wrap it could be useful.

Reply:
Well.... https://github.com/canonical/imagecraft/issues/117 :D